File: scripts/fetch_faces.mjs
#!/usr/bin/env node
/* ============================================================
   Pluck · one-time player-portrait cache
   ------------------------------------------------------------
   SofaScore now 403-blocks hotlinked image requests at runtime,
   so instead of the game calling their API live, we download each
   player's portrait ONCE and serve it locally from /faces/{id}.png.

   Run this on a normal (residential) connection — datacenter / cloud
   IPs are the ones SofaScore blocks:

       node scripts/fetch_faces.mjs

   • Reads every sofascore_id in data/soccer.json (deduped).
   • Fetches one image at a time with a delay, so it never trips
     the rate-limiter. Tune with DELAY_MS, e.g.:
         DELAY_MS=700 node scripts/fetch_faces.mjs
   • Resumable: a player whose faces/{id}.png already exists is
     skipped, so you can stop/rerun freely.
   • Saves the RAW image; the game does the white-background
     flood-fill in the browser (works because the file is now
     same-origin, so the canvas isn't CORS-tainted).

   After it finishes:  git add faces && git commit && git push
   ============================================================ */

import { readFile, mkdir, writeFile, stat } from 'node:fs/promises';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';

const ROOT      = join(dirname(fileURLToPath(import.meta.url)), '..');
const DATA_FILE = join(ROOT, 'data', 'soccer.json');
const FACES_DIR = join(ROOT, 'faces');
const DELAY_MS  = Number(process.env.DELAY_MS || 450);

const HEADERS = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
    'Referer': 'https://www.sofascore.com/',
    'Accept': 'image/avif,image/webp,image/png,image/*,*/*;q=0.8',
};

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));
const fileReady = async (p) => { try { return (await stat(p)).size > 0; } catch { return false; } };

const teams = JSON.parse(await readFile(DATA_FILE, 'utf8'));
const ids = [...new Set(
    teams.flatMap((t) => (t.lineup || []).map((p) => p.sofascore_id).filter(Boolean))
)];

await mkdir(FACES_DIR, { recursive: true });

let fetched = 0, skipped = 0;
const failed = [];

console.log(`Pluck face cache · ${ids.length} unique player IDs · ${DELAY_MS}ms between requests\n`);

for (let i = 0; i < ids.length; i++) {
    const id  = ids[i];
    const out = join(FACES_DIR, `${id}.png`);
    const tag = `[${String(i + 1).padStart(3)}/${ids.length}]`;

    if (await fileReady(out)) { skipped++; continue; }

    const url = `https://api.sofascore.com/api/v1/player/${id}/image`;
    try {
        const res = await fetch(url, { headers: HEADERS });
        const ct  = res.headers.get('content-type') || '';
        if (!res.ok || !ct.startsWith('image/')) {
            failed.push(`${id} — HTTP ${res.status} (${ct || 'no content-type'})`);
            console.log(`${tag} ✗ ${id}  HTTP ${res.status}`);
        } else {
            const buf = Buffer.from(await res.arrayBuffer());
            if (buf.length < 200) {
                failed.push(`${id} — suspiciously small (${buf.length}b)`);
                console.log(`${tag} ✗ ${id}  tiny (${buf.length}b)`);
            } else {
                await writeFile(out, buf);
                fetched++;
                console.log(`${tag} ✓ ${id}.png  (${(buf.length / 1024).toFixed(1)} KB)`);
            }
        }
    } catch (e) {
        failed.push(`${id} — ${e.message}`);
        console.log(`${tag} ✗ ${id}  ${e.message}`);
    }

    await sleep(DELAY_MS);
}

console.log(`\nDone.  fetched=${fetched}  skipped=${skipped}  failed=${failed.length}`);
if (failed.length) {
    console.log('\nFailures (these players keep the monogram fallback in-game):');
    for (const f of failed) console.log('  ' + f);
    console.log('\nIf MANY failed with HTTP 403, your IP is being rate-limited — rerun');
    console.log('later with a bigger delay, e.g.  DELAY_MS=1200 node scripts/fetch_faces.mjs');
    console.log('(it resumes where it left off).');
}

File: scripts/fetch_faces.py
#!/usr/bin/env python3
"""
Pluck - one-time player-portrait cache (your slow-download approach).

SofaScore now 403-blocks hotlinked/runtime image requests, so instead of the
game calling their API live, we download each player's portrait ONCE and serve
it locally from faces/{id}.png.

    python scripts/fetch_faces.py

* Reads every sofascore_id in data/soccer.json (deduped).
* Downloads one image at a time with a delay (default 1.5s) so you stay under
  the rate limit -- tune with the DELAY env var:  DELAY=2.5 python scripts/fetch_faces.py
* Resumable: a player whose faces/{id}.png already exists is skipped.
* Saves the RAW image; the game does the white-background flood-fill in the
  browser (works because the file is now same-origin, so the canvas isn't
  CORS-tainted).

IMPORTANT: run this from a NORMAL/RESIDENTIAL connection. SofaScore blocks
datacenter/cloud IPs (and aggressively rate-limits). If you get a wall of 403s
on request #1, your IP itself is blocked -- try a different network (e.g. a
phone hotspot or VPN); it's resumable, so you can stop/restart freely.

Stdlib only -- no `pip install` needed. (Mirrors the requests+sleep approach.)
"""

import json
import os
import time
import urllib.request
import urllib.error

ROOT      = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
DATA_FILE = os.path.join(ROOT, "data", "soccer.json")
FACES_DIR = os.path.join(ROOT, "faces")
DELAY     = float(os.environ.get("DELAY", "1.5"))   # seconds between requests
URL       = "https://api.sofascore.com/api/v1/player/{}/image"
HEADERS   = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
                  "(KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
    "Referer": "https://www.sofascore.com/",
    "Accept": "image/avif,image/webp,image/png,image/*,*/*;q=0.8",
}

db = json.load(open(DATA_FILE, encoding="utf-8"))
ids = sorted({p["sofascore_id"] for c in db for p in c.get("lineup", []) if p.get("sofascore_id")})
os.makedirs(FACES_DIR, exist_ok=True)

fetched = skipped = 0
failed = []
print(f"{len(ids)} unique player IDs - {DELAY}s between requests\n")

for i, sid in enumerate(ids, 1):
    out = os.path.join(FACES_DIR, f"{sid}.png")
    if os.path.exists(out) and os.path.getsize(out) > 0:
        skipped += 1
        continue
    try:
        req = urllib.request.Request(URL.format(sid), headers=HEADERS)
        with urllib.request.urlopen(req, timeout=20) as r:
            ctype = r.headers.get("Content-Type", "")
            data = r.read()
        if not ctype.startswith("image/") or len(data) < 200:
            failed.append(f"{sid} (got {ctype or '?'} {len(data)}b)")
            print(f"[{i:>3}/{len(ids)}] x {sid}  not-an-image")
        else:
            with open(out, "wb") as f:
                f.write(data)
            fetched += 1
            print(f"[{i:>3}/{len(ids)}] ok {sid}.png  ({len(data)/1024:.1f} KB)")
    except urllib.error.HTTPError as e:
        failed.append(f"{sid} (HTTP {e.code})")
        print(f"[{i:>3}/{len(ids)}] x {sid}  HTTP {e.code}")
    except Exception as e:
        failed.append(f"{sid} ({e})")
        print(f"[{i:>3}/{len(ids)}] x {sid}  {e}")
    time.sleep(DELAY)

print(f"\nDone. fetched={fetched} skipped={skipped} failed={len(failed)}")
if failed:
    print("\nFailed (these keep the monogram fallback in-game):")
    for f in failed:
        print("  " + f)
    print("\nIf MOST failed with HTTP 403, this connection is blocked. Try a")
    print("different network and rerun (it resumes), or use the browser route:")
    print("  scripts/fetch_faces_browser.js")
